packages/hop-node/src/vault/AaveVault.ts

```diff
@@ -0,0 +1,293 @@
+    }
+    const contract = this.getErc20(this.aTokenAddress)
+    return contract.balanceOf(account)
+  }
+
+  async deposit (amount: BigNumber) {
+    const account = await this.signer.getAddress()
+    console.log('account:', account)
+
+    const needsApproval = await this.needsDepositApproval(amount)
+    if (needsApproval) {
+      console.log('needs approval; sending approval tx')
+      const tx = await this.approveDeposit()
+      console.log('approval tx:', tx.hash)
+      await tx.wait()
+    }
+
+    const deadline = Math.floor(Date.now() / 1000) + (60 * 60)
+    let txs: any[] = []
+
+    // available on aave v3 and doesn't work with DAI
+    const canSupplyWithPermit = this.chain === 'arbitrum' && ['USDC', 'USDT'].includes(this.token)
+    if (canSupplyWithPermit) {
+      const dataToSign = await this.pool.signERC20Approval({
+        user: account,
+        reserve: this.tokenAddress,
+        amount: this.formatUnits(amount).toString(),
+        deadline: deadline.toString()
+      })
+
+      const parsedData = JSON.parse(dataToSign)
+      const { domain, types, message: value } = parsedData
+      delete types.EIP712Domain
+      const signature = await this.signer._signTypedData(domain, types, value)
+
+      txs = await this.pool.supplyWithPermit({
+        user: account,
+        reserve: this.tokenAddress,
+        amount: this.formatUnits(amount).toString(),
+        signature,
+        deadline: deadline.toString()
+      })
+    } else {
+      if (this.chain === Chain.Ethereum) {
+        // aave v2
+        txs = await this.pool.deposit({
+          user: account,
+          reserve: this.tokenAddress,
+          amount: this.formatUnits(amount).toString()
+        })
+      } else {
+        // aave v3
+        txs = await this.pool.supply({
+          user: account,
+          reserve: this.tokenAddress,
+          amount: this.formatUnits(amount).toString()
+        })
+      }
+    }
+
+    console.log('aave txs:', txs)
+    console.log('attempting to deposit')
+
+    let tx: any
+    for (const item of txs) {
+      console.log(await item.gas())
+      const txPayload = await item.tx()
+      if (txPayload.value) {
+        txPayload.value = BigNumber.from(txPayload.value).toHexString()
+      }
+      tx = await this.signer.sendTransaction(txPayload)
+      console.log(tx)
+      await tx.wait()
+    }
+
+    return tx
+  }
+
+  async withdraw (amount: BigNumber) {
+    const account = await this.signer.getAddress()
+    console.log('account:', account)
+
+    const needsApproval = await this.needsWithdrawalApproval(amount)
+    if (needsApproval) {
+      console.log('needs approval; sending approval tx')
+      const tx = await this.approveWithdrawal()
+      console.log('approval tx:', tx.hash)
+      await tx.wait()
+    }
+
+    // aave v2 and v3
+    const txs = await this.pool.withdraw({
+      user: account,
+      reserve: this.tokenAddress,
+      amount: this.formatUnits(amount).toString(),
```

Comment:
Is this `amount` in terms of canonical token or a token? If it is a token, `formatUnits` will need to reflect the a token `decimals` (mentioned in another comment but flagging it here as well).

Reply:
The decimals will always be the same for both canonical token and their aave token (ie USDC and aUSDC both have 6 decimals)